src/v2/catalog.rs

```diff
@@ -1,60 +1,61 @@
 use crate::errors::{Error, Result};
 use crate::v2;
-use futures::{self, stream, Future, Stream};
-use reqwest::StatusCode;
-use serde_json;
-
-/// Convenience alias for a stream of `String` repos.
-pub type StreamCatalog = Box<dyn futures::Stream<Item = String, Error = Error>>;
+use async_stream::try_stream;
+use futures::stream::Stream;
+use futures::{self};
+use reqwest::{RequestBuilder, StatusCode};
 
 #[derive(Debug, Default, Deserialize, Serialize)]
 struct Catalog {
     pub repositories: Vec<String>,
 }
 
 impl v2::Client {
-    pub fn get_catalog(&self, paginate: Option<u32>) -> StreamCatalog {
-        let url = {
-            let suffix = if let Some(n) = paginate {
-                format!("?n={}", n)
-            } else {
-                "".to_string()
+    pub fn get_catalog<'a, 'b: 'a>(
+        &'b self,
+        paginate: Option<u32>,
+    ) -> impl Stream<Item = Result<String>> + 'a {
+        try_stream! {
+            let url = {
```

Comment:
Would it be possible to move this "once-init" logic out of the `try_stream!()` block? It doesn't seem to affect/changt with stream iteration, so having it outside of the macro block would improve readability.

Reply:
The `?` operator behaves differently inside the macro block. If I simply move the code outside, the compiler says:

```
error[E0277]: the `?` operator can only be used in  that returns `Result` or `Option` (or another type that implements `std::ops::Try`)
  --> src/v2/catalog.rs:26:13
   |
26 | /             reqwest::Url::parse(&ep).map_err(|e| {
27 | |                 Error::from(format!("failed to parse url from string '{}': {}", ep, e))
28 | |             })?
   | |_______________^ cannot use the `?` operator in  that returns `async_stream::async_stream::AsyncStream<std::result::Result<std::string::String, errors::Error>, impl core::future::future::Future>`
   |
   = help: the trait `std::ops::Try` is not implemented for `async_stream::async_stream::AsyncStream<std::result::Result<std::string::String, errors::Error>, impl core::future::future::Future>`
   = note: required by `std::ops::Try::from_error`
```

I might be able to work around this by manually constructing the stream error value, but I'm not sure if it's worth it. WDYT?

Reply:
Not worth the complexity, I guess. However you can maybe still move most of it out of the macro block, and only keep the `?` inside. Assuming the resulting thing is not uglier. Your call.